### DISCUSSION TITLE
[Feature request] Faster syntax highlighting & autocompletion for larger files (10 000 lines of code +)

### DISCUSSION THREAD

**Discussion by ItsCubeTime** [microsoft/pylance-release#3711] [Enhancement]
Hi, Ive been working on a tool for a couple of months now where most of the code have all gone in a single file (due to some limitations with the framework I use, it have proved to be more work effective than spreading the code across several files as Im the only contributor to the project).

Ive noticed that it takes quite long however for the syntax highlighting (and to some extent also the autocompletion) to update when making changes further down in the file (several seconds even on an i9-12900K, 32gb DDR4 system) - a lot longer than it takes to fire up the script with python and start using it.

If theres any room for speed improvements, I would highly appreciate it.

https://user-images.githubusercontent.com/20190653/164794082-a534e569-4630-4ca3-9ed0-b3f1ee58aa76.mp4

## Demo files
In case a test file would be useful, the `__init__.py` in this zip is the file in question
[slow syntax highlight demo files.zip](https://github.com/microsoft/vscode-python/files/8544045/slow.syntax.highlight.demo.files.zip) (if you want type hints for the `bpy` dependency used in `__init__.py`, you can also install https://github.com/nutti/fake-bpy-module via `pip install fake-bpy-module`, but I suspect it wont be necessary just to benchmark the static analyzer).
The code is 100% written by me, so I have the rights to post it here. Unfortunately this code cant actually run on its own & it also has several dependencies that the language server is likely going to complain about the absence of when testing on another system. If desired I can list all the modules that can be installed via pip

**Comment by erictraut** [comment]
Are you using VS Code with Pylance? I just tried opening your file `__init__.py` with the latest version of Pylance (2022.4.2) using VS Code, and code analysis and syntax highlighting takes about 800ms on my computer. That's longer than typically seen for a Python source file, but it's reasonable given the number of lines of code in this file. You said that it takes "several seconds" on your machine? That implies that you're seeing times in excess of 3000ms?

Pylance is already highly optimized, so it's unlikely that we could make significant performance improvements for analysis of typical code patterns. I don't see anything especially atypical about the code in your file (e.g. highly nested loops, thousands of code flow statements within a single function, or thousands of local variables). Your code even includes type annotations, which reduces the need for the analyzer to infer variable types — a common source of performance issues.

The recommended workaround here is to break your code into multiple modules.

**Comment by heejaechang** [comment]
I am going to take a look where we spend time most of time during semantic colorization.

**Comment by ItsCubeTime** [comment]
> I am going to take a look where we spend time most of time during semantic colorization.

Thank you, much appreciated 💝✨

**Comment by judej** [comment]
Moving this issue to discussion as an enhancement request for comments and upvotes.

**Comment by ItsCubeTime** [comment]
Bump :) Came across https://pypi.org/project/vulkan/ a couple of days ago, results in the same issue when working inside any file that imports its relatively large _vulkan.py file (22.5K lines of code).

![image](https://github.com/microsoft/pylance-release/assets/20190653/e09a62ea-0e48-4d7c-b82a-6bdc759affc7)
Takes something like 30 seconds to a minute or so for autocomplete features to complete on my i9-12900K desktop processor (with 32gb of ram and a couple of TB sized high speed NVMe SSDs) - and it doesn't seem to get any faster the second time you try accessing highlight for an already parsed symbol.

Im attaching my project down here if anyone wants to take a look (prerequisites are `pip install vulkan`, `pip install glfw`, `pip install numpy` & `pip install pyrr`)
[pyvulkan.zip](https://github.com/user-attachments/files/15749149/pyvulkan.zip) (forked from https://github.com/amengede/getIntoGameDev/tree/main/pyvulkan/22%20-%20Indexed%20Rendering/finished )

Video demo: https://youtu.be/JVO0elKshW0

Whats most bothersome, is that once it has "analyzed" the large `_vulkan.py` from top to bottom, it seems to reanalyze it the very next time it needs any information from it once more - even though it may have already fetched that info before.

**Reply by rchiodo** [reply]
Thanks I can repro. My guess is that since vulkan isn't typed, our type evaluation is getting stuck in some really deep tree to figure out return values. That's usually what causes this sort of thing.

Setting up the settings as described here:

https://github.com/microsoft/pylance-release/wiki/Collecting-data-for-an-investigation.#collecting-type-evaluation-related-data-in-the-log

I can see that it's this function that's taking all of the time:

```
2024-06-10 09:14:50.274 [info] (25208) [BG(1)]                       Zt ["_cffi_backend.FFI('vulkan._vul <shortened> " (..vulkan._vulkancache) [4:7]] (28304ms)
```

Not sure why yet. Might just be the size of the line.

**Reply by rchiodo** [reply]
Oh and every time you type, we have to reanalyze all imports as your typing can impact what and how something is imported. That's why there's no cache for this vulcan module.

**Comment by ItsCubeTime** [comment]
Bump :)

Wouldnt it be possible to implement some kind of caching mechanism where it first offers type hints based on what information was available last the files were analyzed (that gets displayed immediately, while reparsing is undergoing)?

Or maybe only reparsing of the current scope? Like if you're typing inside a function, instead of reparsing the entire file, just reparse that function

**Comment by bschnurr** [comment]
see https://github.com/microsoft/pylance-release/wiki/Collecting-data-for-an-investigation.#collecting-type-evaluation-related-data-in-the-log   

turn on these settings and check the language server logs for which expression is taking all the time.
set python.analysis.logLevel to Trace
set python.analysis.logTypeEvaluationTime to true
set python.analysis.typeEvaluationTimeThreshold to 500

might need to generate typestubs

**Reply by ItsCubeTime** [reply]
https://pastebin.com/nBGhr43L
> 2026-06-30 06:23:46.395 [info] [Info  - 6:23:46 AM] (64656) BG: [BG(1)] Long operation: runWithCancellationToken (3908ms)

**Reply by ItsCubeTime** [reply]
You can also benchmark it yourself using the files provided in the original post if you like, its the same file :-) Or the linked vulkan binding project

Edit: Oh and also you may want to pip install: https://github.com/nutti/fake-bpy-module `pip install fake-bpy-module` . Its the most used dependency in the file I provided
